### ISSUE TITLE
Support HTML as a Clipboard format for StyledText

### ISSUE THREAD

**Issue by mihnita** [eclipse-platform/eclipse.platform.swt#292]
The current `StyledText` can copy / paste styled text to clipboard, but only in RTF format.

If you copy a piece of code (with syntax) and paste it in MS Word, or PowerPoint, the colors and font are preserved
(although the background color is lost).
But only plain text is pasted into Google GMail, Docs, Slides, or Yahoo Mail.

**Describe the solution you'd like**

With web-based applications as common as they are, it would be really handy to also support HTML as a clipboard format.
So that one can copy code and paste it in something like Google Slides, for example, while preserving the colors (and font).
The way it works today in PowerPoint.

**Describe alternatives you've considered**

Copy / paste in MS Word works, then copy / paste from MS Word to web-based apps works.
Opening the code in Visual Studio Code and copying from there also works.

But it would be nice to have that working from Eclipse.

**Additional context**

Not much, I think it is pretty obvious.

BUT: I offer to implement this, if you are interested.

Thank you very much,
Mihai

**Comment by mickaelistria** [comment]
Such a contribution would be very welcome!

**Comment by mihnita** [comment]
Hi, 

Sorry, nothing new, but I just to make sure I manage expectations :-)

I did not drop out, but I think this is not something for 2022-09.
My guess is that it is a bit too late for that.
(I also don't have much time, so I would rather focus on the color support in the Console, for now)

But I can share a bit the big picture plan:
* Optional: extract `TextWriter` and `RTFWriter` out of `org.eclipse.swt.custom.StyledText`
* Add an `HTMLWriter`, very similar to the `RTFWriter`
* Change the `StyledText` copy operation to also put the HTML in the clipboard

The HTML I would produce is very "primitive" for best compatibility.
Using `<b>` and `<i>` and `<span style=...>`, no stylesheets or anything fancy.
I've compared the kind of HTML put in the clipboard by all kinds of apps, from (the "ancient") Office 2010 to Google Slides/Docs/GMail.
And that kind of HTML seems to be the norm.

Regards,
Mihai

**Comment by tmssngr** [comment]
Maybe such code can be placed outside the (already large) `StyledText` code?

**Comment by mickaelistria** [comment]
@tmssngr I believe this is what was meant with "Optional: extract TextWriter and RTFWriter out of org.eclipse.swt.custom.StyledText"

**Comment by mihnita** [comment]
Agree with `HTMLWriter` outside `StyledText`.

In fact that was my thinking was to have `HTMLWriter` outside.

And also move `RTFWriter` and `TextWriter` out of `StyledText`.
Right now `RTFWriter` and `TextWriter` are inner classes to `StyledText`.
I think it would be cleaner to make them standalone classes.
They are not public, so that would not be a breaking change.

But this refactoring is not a change that is needed for the html clipboard functionality.
That is why I said "optional".

Mihai

**Comment by mickaelistria** [comment]
@mihnita If you want to start with a refactoring to extract the existing RTFWriter and TextWriter, to familiarize with the code and maybe factorize what seems really worth it, that would be welcome in a preliminary PR.

**Comment by mihnita** [comment]
Thank you.

Right now I am a bit under pressure with a project at work.
I will have a bit more time in a week or two.

And Eclipse is close to release, too late to make it into that. So no rush...
I guess there is some development branch going on for after-release code?
 
Thanks,
Mihai

**Comment by mihnita** [comment]
Since this has a "noteworthy"  label I suppose I will need to prepare an entry for N&N
(https://github.com/eclipse-platform/www.eclipse.org-eclipse-news)

But there is no 4.26 folder yet, so I will put a reminder somewhere.

---

Meantime I will test on other OSes and applications.
Maybe find an MS Office from this century ;-)

M.

**Comment by mickaelistria** [comment]
> Since this has a "noteworthy" label I suppose I will need to prepare an entry for N&N (https://github.com/eclipse-platform/www.eclipse.org-eclipse-news)

Yes, but no need to rush.

> But there is no 4.26 folder yet, so I will put a reminder somewhere.

I'm creating one right now.

**Comment by mickaelistria** [comment]
4.26 folder is ready.
I verified it by copy-pasting from StyledText to GMail, and can now see the styling is preserved.

**Comment by jukzi** [comment]
please fix the regression of adding extra newlines in the htmlText #593

**Comment by jukzi** [comment]
There is an additional newline at the end and every linebreak is formatted as an paragraph instead of linebreak
![image](https://user-images.githubusercontent.com/51790620/225938710-19a4bdc6-a1d2-4d13-90ef-515a60fa21fd.png)

**Comment by mihnita** [comment]
I checked and there are are indeed paragraphs, not line breaks.
But I don't think this is legacy.
It follows exactly the RTF an (almost exactly) the plain text behavior).

I've prepared a Word document with 3 sections, for plain text, RTF, and HTML.
There paragraph markers that already exist are Arial, so it will be clear what comes from clipboard.
Then I've pasted the various formats, with paste special:

![image](https://user-images.githubusercontent.com/1307726/227167702-ee908fb0-6765-4121-a98e-48e02b425cf0.png)

All 3 formats use paragraphs (`<p>`) to separate the lines, not line break (`<br>`)
The RTF and HTML are virtually identical, and there is indeed a paragraph after the `else {` which is missing in plain text.

We can debate if that is correct or not. But is is not a regression, the HTML behave the same as the RTF, which was untouched recently.

---

Now, I would also argue that the extra paragraph is indeed correct, because the editor selection (as shown in the screenshot in [jukzi](https://github.com/jukzi)'s comment) includes a new line.

Here are the results when the EOL after `else {` is not selected:
![image](https://user-images.githubusercontent.com/1307726/227172413-0f87ca12-7a83-4a84-bb32-c2cb719a5ee4.png)

There is indeed a difference between HTML and RTF now (and it is clear what paragraph marker was there and what comes from the clipboard because of the font difference.

So I might agree that his is indeed a bug.
BUT!
There is a reason why is there.

The HTML clipboard contains a `<div>` with several `<p>`s. And I think that MS Words generates that paragraph from the div.
And the div is there to support background colors, which are not supported in RTF:
![image](https://user-images.githubusercontent.com/1307726/227173913-28c1460c-3a51-4a4f-9ab9-7e555357ec36.png)

**Comment by mihnita** [comment]
TLDR: the html I generate is very "old style" (not style sheets, nothing fancy), as from the various apps I've tried many don't support styles, or background colors, and so on.

For example in Word you can do color paragraphs without the div, but Google Slides does not.
The div is also needed to control the line height. And for indents in Google Docs and Slides and 

Here is documented everything I've tried:
https://github.com/eclipse-platform/eclipse.platform.swt/pull/469
(with a link to screenshots before / after: https://mihai-nita.net/eclipse_swt_467/)

The `div` was the only thing that I could find that seem to work everywhere and fixes everything
(background colors, line spacing, indents)

With that price, an extra newline...

**Comment by jukzi** [comment]
The pasted HTML results in an additional gap per paragraph in word (0 when using RTF):
![image](https://user-images.githubusercontent.com/51790620/227217766-decdec55-4b97-4944-ab67-ccca2fd5c0f4.png)

**Comment by mihnita** [comment]
I've spent on this a lot of hours over the week-end.

And yes, there is an extra paragraph in the HTML format.

For line spacing, there is no visual difference between HTML and RTF.
It looks the same as the RTF, and when I inspecting the paragraph properties (same dialog as yours) the spacing before / after is zero.

---

I will still spend some time on it.
But what makes it hard is solving this for everybody.

It is easy to test Outlook vs Word. Both on Windows, both Microsoft, both Office, probably same version.
It looks like even the version of Office matters, as you see different results in line space than what I see.

But this has to work on all OSes, most desktop apps, most web based apps in most popular browsers.

And the copy should not only be from Eclipse code but also any SWT widget.
That means supporting styles that the Eclipse code does not use.

See https://mihai-nita.net/eclipse_clip_html/

You will notice paragraph alignment (left / center / right), bullets, bidi support, justify text or not, enable / disable word wrap, padding (for the whole content), line spacing (from the SWT widget, not fixed), indent, margins.
And this only lists the paragraph attributes, not spans (supperscript, underscore, etc)

---

I also looked at what Word puts in the clipboard.
And it is far from clean, and not at all portable.

This simple code:
```java
if (name != null) {
	System.out.printf("Hello %s!", name);
}
```

Produces 1451 bytes of clipboard data when copied from Eclipse, and 23327 bytes when copied from Word.

And the Word HTML contains "non-standard tricks":
* `<o:p></o:p>` tags from a proprietary namespace (`xmlns:o="urn:schemas-microsoft-com:office:office"`)
* Office specific comments (`)
* Replacing real tabs with several nbsp and wrapped in a `<span style='mso-tab-count:3'>`  (that is for 3 tabs)
  So when you paste that in word (again) the nbsp(s) are in fact replaced by real tabs. But that is a non-portable hack.
* tons of MS styles (MsoNormal)
* tons of proprietary attributes in style (`mso-spacerun`, `mso-tab-count`, in general tons of `mso-` attributes)

It also seems to produce different html depending on what I "grab" in the clipboard.
For plain (ASCII) text it produces fewer proprietary attributes and styles.
Grab some fancy symbol character and it adds `mso-ascii-font-family`, `mso-ascii-theme-font`, `mso-hansi-font-family`, `mso-hansi-theme-font`, `mso-char-type`, `mso-symbol-font-family` and many other attributes.

In all this is what I see in the body part. I've completely ignored non-standard stuff happening in the style definitions.

TLDR: what Word (and Office) produces can't be a good example for what we want to put there.

---

On the good side, I think I have a small improvement that would remove the extra newline **when there was a newline selected in Eclipse**.

But I have to test it on other platforms and applications.

**Comment by laeubi** [comment]
@mihnita honestly HTML "format" was from my opinion always broken (regardless of application) and especially for microsoft as you have discovered now ;-)

So from my side, if we have basic support for this this is totally fine and things like an additional line break a really nothing to worry about...

**Comment by jukzi** [comment]
Using "<span>"'s instead of "<div>" and "<p>" for formatting and only adding "<br>" or "<p>" where needed both Teams and Word don't add additional newlines (but sill increased paragraph hight).

**Comment by mihnita** [comment]
These are narrow solutions that only solve this problem, only for Office, and only for code.

I am looking for something that works everywhere, and anything that SWT StyledText supports.
`<p>` and `<div>` are block elements (similar to paragraph styles in Word), `<span>` is an inline element (similar to character style in Word)
This means that text in a `<span>` can be centered or justified, for example.

Check the [`StyledText`](https://help.eclipse.org/latest/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fswt%2Fcustom%2FStyledText.html) documentation.

There are things that affect the whole widget (and belong in a `div`), for example `setAlignment`, `setBackground`, `set*Margin`, `setIndent`, `setJustify`, `setOrientation`, etc.
Then there are line attributes that work (most of them) on block elements only (`setLineJustify`, `setLineBullet, `setLineSpacing`, 

The things that map to `<span>` are the `StyleRange`(s), (used in `setStyleRanges`)

---

At this point I am ready to bet that the HTML clipboard (as is is now) supports more of the `StyledText` style attributes than the RTF clipboard format.

---

Just give me a some of breathing space and I will prepare something, and test it to work properly everywhere.

Or you can do something your way and I will test it on all platforms and with all the styles (including non-code editor ones) and I will tell you what breaks.

**Comment by mihnita** [comment]
> honestly HTML "format" was from my opinion always broken (regardless of application) and especially for microsoft as you have discovered now ;-)

Well, HTML was not supported at all.
So "working better now" is a pretty low bar :-)
